nixos/modules/services/web-apps/prosody-filer.nix

```diff
@@ -0,0 +1,63 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+
+  cfg = config.services.prosody-filer;
+
+  settingsFormat = pkgs.formats.toml { };
+  configFile = settingsFormat.generate "prosody-filer.toml" cfg.settings;
+in {
+
+  options = {
+    services.prosody-filer = {
+      enable = mkEnableOption "Prosody Filer XMPP upload file server";
+
+      settings = mkOption {
+        description = "Configuration for Prosody Filer.";
+
+        type = types.submodule {
+          freeformType = settingsFormat.type;
+        };
+
+        example = literalExample ''
+          {
+            listenport = "[::1]:5050";
+            secret = "mysecret";
+            storeDir = "/var/lib/upload";
+            uploadSubDir = "upload/";
+          }
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    services.prosody-filer.settings = {
+      listenport = mkDefault "127.0.0.1:5050";
+      uploadSubDir = mkDefault "upload/";
+    };
+
+    users.users.prosody-filer = {
+      group = "prosody-filer";
+      isSystemUser = true;
+      uid = config.ids.uids.prosody-filer;
+    };
+
+    users.groups.prosody-filer = {
+      gid = config.ids.gids.prosody-filer;
+    };
+
+    systemd.services.prosody-filer = {
+      description = "Prosody file upload server";
+      wantedBy = [ "multi-user.target" ];
```

Comment:
maybe an `after = [ "network.target" ];` would also be useful here to avoid spurious start failures when the network isn't fully up yet